### PR TITLE
[CFE-840] Proposes metadata controller for user tag management on cloud resources

### PR DESCRIPTION
**Pull request by TrilokGeer**
The enhancement proposes user tag management controller for day 2 operations. The proposal is scoped for AWS cloud provider resources.

### CODE REVIEW

**Comment by TrilokGeer** [comment]
Link to ADRs : https://github.com/TrilokGeer/metadata-controller/tree/main/docs/decisions

File: enhancements/metadata/user-tag-management.md [hunk 1]
```diff
@@ -0,0 +1,372 @@
...
+ cloud resources created by kubernetes services.
+2. Provide a new custom resource of kubernetes api for specification and status of tags.
+3. Enable continuous monitoring of tags applied and reconcile when there are changes made external to 
+the controller's API.
+4. Provide CLI for operations, cloud service provide authentication and ability to edit custom resource with ease.
+
+### Non-goals
+
+1. Managing metadata which are not related to cloud resource tagging.
+2. Data-related metadata for stored objects and records, for example, S3 bucket vs S3 object.
```
Comment (bparees) [inline enhancements/metadata/user-tag-management.md:63]: can you make this clearer?  I think it's trying to say we'd manage the metadata for things like s3 buckets, but not for things like s3 objects.  (we'd manage the metadata for the cloud resource but not the content within the resource)

i'd also say, assuming i got that right, that it's more than a non-goal.  Non-goals are things you aren't setting out to solve, but take no opinion on whether they happen or not.  But we already know managing metadata on s3 objects is highly problematic because of the tag limit, right?  So i think we should explicitly state that those resource types are off the table for being managed by this controller.

(also if there's another reason to avoid managing metadata on those objects, beyond just the problem of the tag limit constraints, we should state that somewhere.)

File: enhancements/metadata/user-tag-management.md [hunk 2]
```diff
@@ -0,0 +1,372 @@
...
+3. Sub-set of input tags to be applied on selected cloud resource.
+4. Managing conflict with cloud service provider tagging policies.
+5. Preservation of pre-existing tags or tags added using external tools.
```
Comment (bparees) [inline enhancements/metadata/user-tag-management.md:66]: i think we've previously said we wouldn't touch existing tags on resources for fear of removing a tag a customer added manually.  Is this moving away from that position?

File: enhancements/metadata/user-tag-management.md [hunk 3]
```diff
@@ -0,0 +1,372 @@
...
+6. Regular expression or wildcard characters for classifier tag.
+7. Retry mechanism for failed operation during reconciliation.
```
Comment (bparees) [inline enhancements/metadata/user-tag-management.md:68]: why is this a non-goal?

File: enhancements/metadata/user-tag-management.md [hunk 4]
```diff
@@ -0,0 +1,372 @@
...
+7. Add/update tag workflow is initiated.
+
+#### Add/Update tags
+
+1. User creates `AWSMetadata` object with tag list details.
+2. User creates `CloudMetadata` configuration with controller configuration.
+3. Controller updates ready condition to false in `MetadataStatus.status`.
+4. Controller validates tag entries from `AWSMetadata.spec.resourcetags` and populates a new list of tags to be added to cloud resource.
+5. Controller trims the tag list as per maximum limit policy specific to cloud resource.
+6. Controller performs lexicographic sorting.
```
Comment (bparees) [inline enhancements/metadata/user-tag-management.md:116]: needs to sort before trimming, to ensure deterministic set of tags to apply.

File: enhancements/metadata/user-tag-management.md [hunk 5]
```diff
@@ -0,0 +1,372 @@
...
+7. Controller replaces the tag list on the cloud resource.
```
Comment (bparees) [inline enhancements/metadata/user-tag-management.md:117]: this says replaces, but later you say it adds them to the existing tags.  I think you actually intend to do the latter, so this step should be rewritten.

File: enhancements/metadata/user-tag-management.md [hunk 6]
```diff
@@ -0,0 +1,372 @@
...
+
+#### Handling of pre-existing tags or tags added using other tools
+
+In case of add operation, controller appends tags to existing tag list. If there is any tag pre-existing with same key name,
+there will be an error reported for the operation. While, in case of update operation, the pre-existing tag is updated. There is
+no distinction made between pre-existing tags and tags in `AWSMetadataSpec` for reconciliation. The active tag list is listed in the status.
+
+#### Classifier tag behaviour
+
+Classifier tags are used to enable controller to get applicable cloud resources to manage tags. Classifier tags are mandatory and should
```
Comment (bparees) [inline enhancements/metadata/user-tag-management.md:139]: I think you mean "classifier tags are used to identify cloud resources for which the controller should manage the tags"?   But it was a bit hard to follow.

File: enhancements/metadata/user-tag-management.md [hunk 7]
```diff
@@ -0,0 +1,372 @@
...
+be added to cloud resource. Controller does not support any wildcard characters in classifier strings.
+When multiple classifiers are used, a logical OR condition is applied for the classifiers.
+
+#### Reconciliation
+
+Reconciliation of tags is ignored when ready condition is set to false at `MetadataStatus.status`. Reconciliation of tags is based on active tag list in `AWSMetadata.status.resourcetags`. On deletion of AWSMetadataRef
```
Comment (bparees) [inline enhancements/metadata/user-tag-management.md:145]: metadatastatus.status on what resource?  and when would that ready condition be set to false?

Below you say it would be set to false when there's no AWSMetadataRef, but i don't think that's what i'd expect....the controller is still ready(assuming that's what the condition refers to), it just has nothing to do.

File: enhancements/metadata/user-tag-management.md [hunk 8]
```diff
@@ -0,0 +1,372 @@
...
+or invalid reference, controller does not perform any reconciliation. Ready condition is set to false at `MetadataStatus.status`.
```
Comment (bparees) [inline enhancements/metadata/user-tag-management.md:146]: invalid reference from what to what?

File: enhancements/metadata/user-tag-management.md [hunk 9]
```diff
@@ -0,0 +1,372 @@
...
+
+## Summary
+
+Tagging cloud resources enable users to perform administrative operations like, organize the resources, 
+apply security policies, optimize operations, etc. The cloud resources can be tagged using cloud service provider tools, 
+kubernetes controllers that create resources via cloud service provider api and other open source tools. A reconciliation 
+using a generic metadata controller helps to keep the tags synchronized for cloud resources. 
+
+In the following proposal, the controller supports mainly 
+1. Day 2 operations and tag management for cloud resources selected using classifiers.
```
Comment (bparees) [inline enhancements/metadata/user-tag-management.md:16]: who applies the classifier tags?  The operator creating the resource?  the admin?  other?

**Review by bparees — COMMENTED** [review]
i'm struggling w/ the use case for this (it feels a bit tangential to OCP's core purpose because it seems like a generic tool for managing tags on specific cloud resources identified by...an administrator?), and also how it would interact w/ the existing logic that's been added to our installer+operators to put tags on resources we manage.

Would we migrate existing operators that contain tag management code, to instead mark their resources with a "classifier" and rely on this controller to put the cluster-scope-defined tags on the resource?

do any of those existing operators put their own tags on the resources that aren't cluster-scoped/common tags?

@sdodson @JoelSpeed this feels most closely aligned w/ cluster infrastructure so tagging you in as well for your thoughts.

File: enhancements/metadata/user-tag-management.md (hunk 9 above)
Comment (bparees) [reply enhancements/metadata/user-tag-management.md:16]: > The classifier tag is added by admin. It can be added by installer or operators creating the resource (existing flow of adding tags on cloud resource creation) or using external tool on the cloud resource.

so the admin would provide the classifier tag as an additional tag to be applied, to the installer.

the installer/operators would apply that tag during resource creation

then on day 2 the admin would be able to leverage this controller to define additional tags to be applied to the resources that got the classifier tag?

something about that flow feels awkward ("i want to tag resources, so first i have to tag resources")

**Comment by TrilokGeer** [comment]
>

**Comment by TrilokGeer** [comment]
> so is there a user/customer/SD/PM request behind this?

The link heirarchy is lengthy :) . https://issues.redhat.com/browse/RFE-3677 presents the usecase.

**Comment by bparees** [comment]
> The link heirarchy is lengthy :) . https://issues.redhat.com/browse/RFE-3677 presents the usecase.

thanks.  From that ticket:

> 3. Why does the customer need this? (List the business requirements here)
For additional billing and metering insight within the AWS platform

Not clear to me why just tagging all the resources openshift creates, with some openshift key and cluster id value, isn't sufficient?

Does the customer need to be able to customize the tag, or does the customer just need their to be a unique tag that makes it clear that the resource is associated to an openshift cluster, and which cluster?

i'm pushing back on this a bit because there are a lot of gotchas around managing resource tags, as we've already learned from what we've done so far (mostly around tag limits, but also the implications for conflicts/updates, user editing, which resources get their tags managed and which don't, etc.).  So i think we want to be really sure we need to build something this generic before we do so.

cc @heheffne

File: enhancements/metadata/user-tag-management.md [hunk 10]
```diff
@@ -0,0 +1,396 @@
...
+1. POLICY_APPLY_PARTIAL - Allows input tag list applied to be partial on cloud resource. 
+2. POLICY_APPLY_STRICT - Applies input tag list strictly in complete.
+
+### Opcodes
+
+Controller supports different opcodes for actions to add, update and delete tags. Following are the supported opcodes
+1. "add" - Controller adds tags to the cloud resources. In case of an existing entry for the tag, controller reports failure.
+2. "update" - Controller updates existing tags for the cloud resources. If there is no existing entry, controller reports failure.
+3. "delete" - Controller removes tags from the cloud resources. In case of specific errors, user intervention may be required to remove tags manually
+from the cloud resources.
```
Comment (jsafrane) [inline enhancements/metadata/user-tag-management.md:91]: OCP APIs are usually declarative and describe the desired state, not how to reach it.

File: enhancements/metadata/user-tag-management.md [hunk 11]
```diff
@@ -0,0 +1,396 @@
...
+4. Controller gets the active tag list, removes requested tags from the list and populates a new list of tags.
+5. Controller replaces the tag list on the cloud resource.
+6. Controller updates the tag list to `AWSMetadata.status`.
+7. Controller updates ready condition to true in `MetadataStatus.status`.
+
+#### Handling of pre-existing tags or tags added using other tools
+
+In case of add operation, controller appends tags to existing tag list. If there is any tag pre-existing with same key name,
+there will be an error reported for the operation. While, in case of update operation, the pre-existing tag is updated. There is
+no distinction made between pre-existing tags and tags in `AWSMetadataSpec` for reconciliation. The active tag list is listed in the status.
```
Comment (jsafrane) [inline enhancements/metadata/user-tag-management.md:130]: > there will be an error reported for the operation

How are these errors reported? Consider a large cluster, with 100 VMs, >100 volumes, >1000 volume snapshots.

File: enhancements/metadata/user-tag-management.md (hunk 10 above)
Comment (JoelSpeed) [reply enhancements/metadata/user-tag-management.md:91]: Typically yes, i would have expected this to have a list of tags in the spec, then track the list of tags applied in the status. You could have an option in the spec to determine how to handle tags that are present in status but not in spec. Either you would orphan them (remove them from status without doing anything) or prune them (remove them from the cloud before removing them from status).

With operations, you then need to define a semantic for, "this operation has completed". And what does that mean if I were to re-add it to the cloud? Does the user then expect that it is re-removed? Then it's not an action any more IMO

File: enhancements/metadata/user-tag-management.md [hunk 12]
```diff
@@ -0,0 +1,396 @@
...
+2. Controller starts to list/watch custom resource.
+3. User creates `AWSMetadata` object with tag list details.
+4. User creates `CloudMetadata` configuration with controller configuration.
+5. Based on the configuration, controller queries and lists all resources based on classifier tag.
+6. Creates list of existing tags on cloud resource.
+7. Add/update tag workflow is initiated.
+
+#### Add/Update tags
+
+1. User creates `AWSMetadata` object with tag list details.
```
Comment (JoelSpeed) [inline enhancements/metadata/user-tag-management.md:106]: Wonder, have you considered the idea of having each tag represented as a single CR, the tag lifecycle then being tied to just a single CR instance, then the removal of the CR from the Kube cluster would represent either the removal from the cloud or orphaning of the tag based on some removal policy, might be cleaner than a singleton CR that houses a number of tags

File: enhancements/metadata/user-tag-management.md [hunk 13]
```diff
@@ -0,0 +1,396 @@
...
+3. Controller updates ready condition to false in `MetadataStatus.status`.
+4. Controller gets the active tag list, removes requested tags from the list and populates a new list of tags.
+5. Controller replaces the tag list on the cloud resource.
+6. Controller updates the tag list to `AWSMetadata.status`.
+7. Controller updates ready condition to true in `MetadataStatus.status`.
+
+#### Handling of pre-existing tags or tags added using other tools
+
+In case of add operation, controller appends tags to existing tag list. If there is any tag pre-existing with same key name,
+there will be an error reported for the operation. While, in case of update operation, the pre-existing tag is updated. There is
```
Comment (JoelSpeed) [inline enhancements/metadata/user-tag-management.md:129]: This isn't very Kube native, typically you define a desired state and the controller moves the state of the world to this.

If I want to add a new tag, what happens? I have to add it to the spec, then move it to an update operation once it's been added? 

Would be better to just have a list of desired tags that the controllers handle

File: enhancements/metadata/user-tag-management.md [hunk 14]
```diff
@@ -0,0 +1,396 @@
...
+no distinction made between pre-existing tags and tags in `AWSMetadataSpec` for reconciliation. The active tag list is listed in the status.
+
+#### Classifier tag behaviour
+
+Classifier tags are used to identify cloud resources for which the controller should manage the tags. Classifier tags are mandatory and should
+be added to cloud resource. Controller does not support any wildcard characters in classifier strings.
+When multiple classifiers are used, a logical OR condition is applied for the classifiers.
```
Comment (JoelSpeed) [inline enhancements/metadata/user-tag-management.md:136]: What's the motivation behind an `OR `condition? This doesn't allow users to apply tags to any subset of resources which sounds like it could be desirable?

File: enhancements/metadata/user-tag-management.md [hunk 15]
```diff
@@ -0,0 +1,396 @@
...
+
+#### Reconciliation
+
+Reconciliation of tags is based on active tag list in `AWSMetadata.status.resourcetags`. Reconciliation of tags is ignored when ready condition is set to false at `CloudMetadata.MetadataStatus.status`. On deletion of AWSMetadataRef
+there is no reconciliation of tags. If AWSMetadataRef refers to non-existing `AWSMetadata` object , controller fails with error condition and updates `CloudMetadata.MetadataStatus.status` to false.
+
+#### No cloud provider configured
```
Comment (JoelSpeed) [inline enhancements/metadata/user-tag-management.md:143]: How will this work for other clouds, eg Azure?

File: enhancements/metadata/user-tag-management.md [hunk 16]
```diff
@@ -0,0 +1,396 @@
...
+
+When there is no cloud provider object configured, there will be no operation by default.
+
+#### Deleting classifier tag
+
+Controller does not delete, update or reconcile classifier tags. Controller refers them as read-only tags.
```
Comment (JoelSpeed) [inline enhancements/metadata/user-tag-management.md:149]: What is the impact of using a classifier tag, since this now uses up 1 out of the limited number of tags on a cloud, eg, IIRC S3 buckets can only have 10 tags, using one as a classifier is limiting that further, has there been any investigation into what the impact might be?

File: enhancements/metadata/user-tag-management.md [hunk 17]
```diff
@@ -0,0 +1,396 @@
...
+  status: true
+  reason: "none"
+  message: "tags are applied successfully to cloud resources"
+  resourcetags:
+    "env": "test"
+    "centre": "eng"
+```
+### GO language structures
+```go
+type CloudMetadata struct {
```
Comment (JoelSpeed) [inline enhancements/metadata/user-tag-management.md:203]: I think it would be useful to have example of multiple clouds. In the past where we've tried to include multiple clouds into a single project (eg MAPI) it's lead to some pretty awkward decisions API wise. It would be good to make sure to avoid those mistakes again. Having examples of how this would be expanded to a second cloud would aid in that discussion IMO

File: enhancements/metadata/user-tag-management.md [hunk 18]
```diff
@@ -0,0 +1,396 @@
...
+### Risks and Mitigations
+
+1. Any user with permission to edit custom resource can influence tags on cloud resources with wide-scoped classifiers.
+
+### Drawbacks
+
+1. User intervention is required in some cases for failure resolution of opcode driven operations.
+2. HA is not supported for the controller.
+3. Override of global configurations specific to cloud provider are not supported.
+4. Pre-existing tags or tags added using external tools on cloud resources can be updated using the controller.
```
Comment (JoelSpeed) [inline enhancements/metadata/user-tag-management.md:331]: Is this a drawback? IMO this is part of enforcing policy, I think that's what this project is creating right, a tag policy for the clusters, making sure it's correct seems important to me, even if that means overwriting external changes

File: enhancements/metadata/user-tag-management.md [hunk 19]
```diff
@@ -0,0 +1,396 @@
...
+5. Cloud provider tag policies will conflict with operations supported by controller. A review of the policies is required
```
Comment (JoelSpeed) [inline enhancements/metadata/user-tag-management.md:332]: I would like to understand how this would interact with things like MAPI which already manage tags on cloud provider resources

**Review by JoelSpeed — COMMENTED** [review]
I think I'd like this enhancement to explain more about how this controller will interact with different cloud resources. For example, on AWS there are a huge number of different resources that can be tagged, how will this controller handle all of these, and why is it better to handle them centrally like this, vs having individual, more targeted controllers handle them? (eg we already have tagging in MAPI and storage operators for the resources they care about?)

I worry that having a generic tagging controller like this may be too big of a scope given the number of different resources available to be tagged across a single cloud let alone many clouds

**Comment by dhellmann** [comment]
(automated message) This pull request is closed with lifecycle/rotten. It does not appear to be linked to a valid Jira ticket. Should the PR be reopened, updated, and merged? If not, removing the lifecycle/rotten label will tell this bot to ignore it in the future.

**Comment by bchandra-ocp** [comment]
> > 3. Why does the customer need this? (List the business requirements here)
> >    For additional billing and metering insight within the AWS platform
> 
> Not clear to me why just tagging all the resources openshift creates, with some openshift key and cluster id value, isn't sufficient?
Does OCP cluster have a tag that is the cluster-id? Can you please share documentation on how to do that? 

> Does the customer need to be able to customize the tag, or does the customer just need their to be a unique tag that makes it clear that the resource is associated to an Openshift cluster, and which cluster?

On AWS resources, customers can remove and add tags when their resource management for cost, permissions or any governance change. A simple example is ability to change the charge-back to different service or team.
